src/libostree/ostree-repo.c

```diff
@@ -156,6 +156,513 @@ G_DEFINE_TYPE (OstreeRepo, ostree_repo, G_TYPE_OBJECT)
+           (cur_state == LOCK_EX) ? "exclusive" : "shared",
+           stack_len);
+  if (stack_len > 1)
+    {
+      int next_state = GPOINTER_TO_INT (g_queue_peek_nth (lock->stack, 1));
+
+      /* Drop back to the previous lock state if it differs */
+      if (next_state != cur_state)
+        {
+          /* We should never drop from shared to exclusive */
+          g_return_val_if_fail (next_state == LOCK_SH, FALSE);
+          g_debug ("Returning lock state to shared");
+          if (!do_repo_lock (lock->fd, next_state | flags))
+            return glnx_throw_errno_prefix (error,
+                                            "Setting repo lock to shared failed");
+        }
+      else
+        g_debug ("Maintaining lock state as %s",
+                 (next_state == LOCK_EX) ? "exclusive" : "shared");
+    }
+  else
+    {
+      /* Lock stack will be empty, unlock */
+      g_debug ("Unlocking repo");
+      if (!do_repo_unlock (lock->fd, flags))
+        return glnx_throw_errno_prefix (error, "Unlocking repo failed");
+    }
+
+  g_queue_pop_head (lock->stack);
+
+  return TRUE;
+}
+
+/**
+ * ostree_repo_lock_push
+ * @self: a #OstreeRepo
+ * @lock_type: the type of lock to acquire
+ * @cancellable: a #GCancellable
+ * @error: a #GError
+ *
+ * Takes a lock on the repository and adds it to the lock stack. If @lock_type
+ * is %OSTREE_REPO_LOCK_SHARED, a shared lock is taken. If @lock_type is
+ * %OSTREE_REPO_LOCK_EXCLUSIVE, an exclusive lock is taken. shared lock is
+ * used. The actual lock state is only changed when locking a previously
+ * unlocked repository or upgrading the lock from shared to exclusive. If the
+ * requested lock state is unchanged or would represent a downgrade (exclusive
+ * to shared), the lock state is not changed and the stack is simply updated.
+ *
+ * ostree_repo_lock_push() waits for the lock depending on the repository's
+ * lock-timeout configuration. When lock-timeout is -1, a blocking lock is
+ * attempted. Otherwise, the lock is taken non-blocking and
+ * ostree_repo_lock_push() will sleep synchronously up to lock-timeout seconds
+ * attempting to acquire the lock. If the lock cannot be acquired within the
+ * timeout, a %G_IO_ERROR_WOULD_BLOCK error is returned.
+ *
+ * If @self is not writable by the user, then no locking is attempted and
+ * %TRUE is returned.
+ *
+ * Returns: %TRUE on success, otherwise %FALSE with @error set
+ * Since: 2017.13
+ */
+gboolean
+ostree_repo_lock_push (OstreeRepo          *self,
+                       OstreeRepoLockType   lock_type,
+                       GCancellable        *cancellable,
+                       GError             **error)
+{
+  g_return_val_if_fail (self != NULL, FALSE);
+  g_return_val_if_fail (self->inited, FALSE);
+
+  if (!self->writable)
+    return TRUE;
+
+  if (self->lock_timeout_seconds == -1)
+    {
+      g_debug ("Pushing lock blocking");
+      return push_repo_lock (self, lock_type, TRUE, error);
+    }
+  else
```

Comment:
Maybe add a `g_assert (self->lock_timeout_seconds >= 0)` here.

Reply:
Since any setter for `self->lock_timeout_seconds` already checks that it's `>= -1`, I decided not to bother here. But I could add the assert to make sure we don't mess up somewhere.

Reply:
I find such assertions help to document the code, and also help keep static analysis tools on the straight and narrow.

Reply:
Fixed in #1343. I put an assert for `>= -1` above the entire block.